cmd/openshift-install/graph.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/awalterschulze/gographviz"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests"
+	"github.com/openshift/installer/pkg/asset/metadata"
+)
+
+var (
+	graphOpts struct {
+		outputFile string
+	}
+)
+
+func newGraphCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "graph",
+		Short: "Outputs the internal dependency graph for installer",
+		Long:  "",
+		RunE:  runGraphCmd,
+	}
+	cmd.PersistentFlags().StringVar(&graphOpts.outputFile, "output-file", "", "file where the graph is written, if empty prints the graph to Stdout.")
+	return cmd
+}
+
+func runGraphCmd(cmd *cobra.Command, args []string) error {
+	g := gographviz.NewGraph()
+	g.SetName("G")
+	g.SetDir(true)
+	g.SetStrict(true)
+	targets := []struct {
+		name string
+		deps []asset.Asset
+	}{{
+		name: `"Install Config"`,
+		deps: []asset.Asset{&installconfig.InstallConfig{}},
+	}, {
+		name: `"Manifests"`,
+		deps: []asset.Asset{&manifests.Manifests{}, &manifests.Tectonic{}},
+	}, {
+		name: `"Ignition Configs"`,
+		deps: []asset.Asset{&bootstrap.Bootstrap{}, &machine.Master{}, &machine.Worker{}},
+	}, {
+		name: `"Cluster"`,
+		deps: []asset.Asset{&cluster.TerraformVariables{}, &kubeconfig.Admin{}, &cluster.Cluster{}, &metadata.Metadata{}},
+	}}
+	targetNodeAttr := map[string]string{
+		string(gographviz.Shape): "box",
+		string(gographviz.Style): "filled",
+	}
+	for _, target := range targets {
+		g.AddNode("G", target.name, targetNodeAttr)
+		for _, dep := range target.deps {
+			addEdge(g, target.name, dep)
+		}
+	}
+
+	out := os.Stdout
+	if graphOpts.outputFile != "" {
+		f, err := os.Create(graphOpts.outputFile)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		out = f
+	}
+
+	if _, err := io.WriteString(out, g.String()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addEdge(g *gographviz.Graph, parent string, asset asset.Asset) {
+	elem := reflect.TypeOf(asset).Elem()
+	name := strings.TrimPrefix(fmt.Sprintf(`"%s/%s"`, elem.PkgPath(), elem.Name()), "github.com/openshift/installer/")
```

Comment:
Do we need `PkgPath`?  `elem.Name` seems fairly unique on its own.

Reply:
It adds context if anybody wants to find the asset.